### DISCUSSION TITLE
How to use different Evaluation Metrics per Model in AutoGluon TimeSeries

### DISCUSSION THREAD

**Discussion by buesve** [autogluon/autogluon#4916] [Q&A]
Hello AutoGluon Team and Community,

I am working with AutoGluon TimeSeries and would like to use different evaluation metrics for different models.

I recently discovered the [TILDE-Q loss function in a new research paper](https://arxiv.org/abs/2210.15050) and have successfully implemented it as a [custom eval_metric](https://auto.gluon.ai/dev/tutorials/timeseries/forecasting-metrics.html#custom-forecast-metrics) in AutoGluon Timeseries. According to the literature, TILDE-Q should only be applied to Deep Learning models because it relies on gradient-based optimization.

However, I also want to train and evaluate classical models and, most importantly, the ensemble using SMAPE.

**Challenge:**
So far, I have not found a way to assign different evaluation metrics to different models in AutoGluon TimeSeries.

**Potential Workarounds I Considered:**

**1. Custom "Wrapper" for Evaluation Metric**

- I could create a custom eval_metric that identifies the model name e.g. via **kwargs and dynamically switches between TILDE-Q and SMAPE implementations.

- Problem: I haven't figured out how to access the model name within the custom eval_metric function.

**2. Two Separate TimeSeriesPredictor Instances**

- One predictor trains only the Deep Learning models with TILDE-Q.

- The other predictor trains classical models with SMAPE and handles the ensemble of both predictors.

- Problem: I couldn’t find any documentation on how to combine models from different TimeSeriesPredictor runs into a single ensemble.

**Questions:**
- Is there a built-in way to assign different evaluation metrics to different models in AutoGluon?

- If not, which of the two workarounds do you recommend?

- If neither is feasible, are there any alternative approaches?

**Thanks a lot for your help!**

**Comment by shchur** [comment]
Hi @buesve, as you correctly pointed out, the `eval_metric` in the `TimeSeriesPredictor` is used primarily for model selection / ensembling. If you would like to modify the training loss used by DL models based on GluonTs, you need to use the `distr_output` hyperparameter of these models ([docs](https://auto.gluon.ai/stable/tutorials/timeseries/forecasting-model-zoo.html#deep-learning-models)). The first option is to use one of the available `Output` classes from GluonTS:
```python
from gluonts.torch.distributions import NormalOutput
from autogluon.timeseries import TimeSeriesPredictor

predictor = TimeSeriesPredictor(...).fit(..., hyperparameters={"PatchTST": {"distr_output": NormalOutput()}})
```

Alternatively, you can define your own output class by inheriting from the [`Output` class in GluonTS](https://github.com/awslabs/gluonts/blob/12089c7e626dd7062684bdfa82faac32da2a9ec8/src/gluonts/torch/distributions/output.py#L63-L144). 

```python
from gluonts.torch.distributions import Output
from autogluon.timeseries import TimeSeriesPredictor

class CustomOutput(Output):
    # define all methods & attributes marked with NotImplementedError in the base class Output
    def loss(...):
         ...
    
predictor = TimeSeriesPredictor(...).fit(..., hyperparameters={"PatchTST": {"distr_output": CustomOutput()}})
```

Note that the TFT model currently does not custom `distr_output` values because of the bug that was fixed on the `master` branch but will only be released in v1.3 (#4899).
